File: vb.net/FftGuitarTuner/MainForm.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Runtime.InteropServices
Imports SoundCapture

Public Partial Class MainForm
	Inherits Form
	Private m_isListenning As Boolean = False

	Public ReadOnly Property IsListenning() As Boolean
		Get
			Return m_isListenning
		End Get
	End Property

	Public Sub New()
		InitializeComponent()
	End Sub

	Private Sub closeButton_Click(sender As Object, e As EventArgs)
		Close()
	End Sub

	Private frequencyInfoSource As FrequencyInfoSource

	Private Sub StopListenning()
		m_isListenning = False
		frequencyInfoSource.[Stop]()
		RemoveHandler frequencyInfoSource.FrequencyDetected, New EventHandler(Of FrequencyDetectedEventArgs)(AddressOf frequencyInfoSource_FrequencyDetected)
		frequencyInfoSource = Nothing
	End Sub

	Private Sub StartListenning(device As SoundCaptureDevice)
		m_isListenning = True
		frequencyInfoSource = New SoundFrequencyInfoSource(device)
		AddHandler frequencyInfoSource.FrequencyDetected, New EventHandler(Of FrequencyDetectedEventArgs)(AddressOf frequencyInfoSource_FrequencyDetected)
		frequencyInfoSource.Listen()
	End Sub

	Private Sub frequencyInfoSource_FrequencyDetected(sender As Object, e As FrequencyDetectedEventArgs)
		If InvokeRequired Then
			BeginInvoke(New EventHandler(Of FrequencyDetectedEventArgs)(AddressOf frequencyInfoSource_FrequencyDetected), sender, e)
		Else
			UpdateFrequecyDisplays(e.Frequency)
		End If
	End Sub

	Private Sub UpdateFrequecyDisplays(frequency As Double)
		If frequency > 0 Then
			frequenciesScale1.SignalDetected = True
			frequenciesScale1.CurrentFrequency = frequency

			frequencyTextBox.Enabled = True
			frequencyTextBox.Text = frequency.ToString("f3")

			Dim closestFrequency As Double
			Dim noteName As String = Nothing
			FindClosestNote(frequency, closestFrequency, noteName)
			closeFrequencyTextBox.Enabled = True
			closeFrequencyTextBox.Text = closestFrequency.ToString("f3")
			noteNameTextBox.Enabled = True
			noteNameTextBox.Text = noteName
		Else
			frequenciesScale1.SignalDetected = False

			frequencyTextBox.Enabled = False
			closeFrequencyTextBox.Enabled = False
			noteNameTextBox.Enabled = False
		End If

	End Sub

	Shared NoteNames As String() = {"A", "A#", "B/H", "C", "C#", "D", _
		"D#", "E", "F", "F#", "G", "G#"}
	Shared ToneStep As Double = Math.Pow(2, 1.0 / 12)

	Private Sub FindClosestNote(frequency As Double, _
		<Out> ByRef closestFrequency As Double, <Out> ByRef noteName As String)
		Const  AFrequency As Double = 440.0
		Const  ToneIndexOffsetToPositives As Integer = 120

		Dim toneIndex As Integer = CInt(Math.Truncate(Math.Round(Math.Log(frequency / AFrequency, ToneStep))))
		noteName = NoteNames((ToneIndexOffsetToPositives + toneIndex) Mod NoteNames.Length)
		closestFrequency = Math.Pow(ToneStep, toneIndex) * AFrequency
	End Sub

	Private Sub listenButton_Click(sender As Object, e As EventArgs)
		Dim device As SoundCaptureDevice = Nothing
		Using form As New SelectDeviceForm()
			If form.ShowDialog() = DialogResult.OK Then
				device = form.SelectedDevice
			End If
		End Using

		If device IsNot Nothing Then
			StartListenning(device)
			UpdateListenStopButtons()
		End If
	End Sub

	Private Sub UpdateListenStopButtons()
		listenButton.Enabled = Not m_isListenning
		stopButton.Enabled = m_isListenning
	End Sub

	Private Sub stopButton_Click(sender As Object, e As EventArgs)
		StopListenning()
		UpdateListenStopButtons()
	End Sub

	Private Sub MainForm_FormClosing(sender As Object, e As FormClosingEventArgs)
		If IsListenning Then
			StopListenning()
		End If
	End Sub
End Class

File: cs/FftGuitarTuner/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using SoundCapture;

namespace FftGuitarTuner
{
    public partial class MainForm : Form
    {
        bool isListenning = false;

        public bool IsListenning
        {
            get { return isListenning; }
        }

        public MainForm()
        {
            InitializeComponent();
        }

        private void closeButton_Click(object sender, EventArgs e)
        {
            Close();
        }

        FrequencyInfoSource frequencyInfoSource;

        private void StopListenning()
        {
            isListenning = false;
            frequencyInfoSource.Stop();
            frequencyInfoSource.FrequencyDetected -= new EventHandler<FrequencyDetectedEventArgs>(frequencyInfoSource_FrequencyDetected);
            frequencyInfoSource = null;
        }

        private void StartListenning(SoundCaptureDevice device)
        {
            isListenning = true;
            frequencyInfoSource = new SoundFrequencyInfoSource(device);
            frequencyInfoSource.FrequencyDetected += new EventHandler<FrequencyDetectedEventArgs>(frequencyInfoSource_FrequencyDetected);
            frequencyInfoSource.Listen();
        }

        void frequencyInfoSource_FrequencyDetected(object sender, FrequencyDetectedEventArgs e)
        {
            if (InvokeRequired)
            {
                BeginInvoke(new EventHandler<FrequencyDetectedEventArgs>(frequencyInfoSource_FrequencyDetected), sender, e);
            }
            else
            {
                UpdateFrequecyDisplays(e.Frequency);
            }
        }

        private void UpdateFrequecyDisplays(double frequency)
        {
            if (frequency > 0)
            {
                frequenciesScale1.SignalDetected = true;
                frequenciesScale1.CurrentFrequency = frequency;

                frequencyTextBox.Enabled = true;
                frequencyTextBox.Text = frequency.ToString("f3");

                double closestFrequency;
                string noteName;
                FindClosestNote(frequency, out closestFrequency, out noteName);
                closeFrequencyTextBox.Enabled = true;
                closeFrequencyTextBox.Text = closestFrequency.ToString("f3");
                noteNameTextBox.Enabled = true;
                noteNameTextBox.Text = noteName;
            }
            else
            {
                frequenciesScale1.SignalDetected = false;

                frequencyTextBox.Enabled = false;
                closeFrequencyTextBox.Enabled = false;
                noteNameTextBox.Enabled = false;
            }

        }

        static string[] NoteNames = {"A", "A#", "B/H", "C", "C#", "D", "D#", "E", "F", "F#",  "G",  "G#" };
        static double ToneStep = Math.Pow(2, 1.0 / 12);

        private void FindClosestNote(double frequency, out double closestFrequency, out string noteName)
        {
            const double AFrequency = 440.0;
            const int ToneIndexOffsetToPositives = 120;

            int toneIndex = (int)Math.Round( Math.Log(frequency / AFrequency, ToneStep) );
            noteName = NoteNames[(ToneIndexOffsetToPositives + toneIndex) % NoteNames.Length];
            closestFrequency = Math.Pow(ToneStep, toneIndex) * AFrequency;
        }

        private void listenButton_Click(object sender, EventArgs e)
        {
            SoundCaptureDevice device = null;
            using (SelectDeviceForm form = new SelectDeviceForm())
            {
                if (form.ShowDialog() == DialogResult.OK)
                {
                    device = form.SelectedDevice;
                }
            }

            if (device != null)
            {
                StartListenning(device);
                UpdateListenStopButtons();
            }
        }

        private void UpdateListenStopButtons()
        {
            listenButton.Enabled = !isListenning;
            stopButton.Enabled = isListenning;
        }

        private void stopButton_Click(object sender, EventArgs e)
        {
            StopListenning();
            UpdateListenStopButtons();
        }

        private void MainForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (IsListenning)
            {
                StopListenning();
            }
        }
    }
}
